File: Lw5Sharp/Lw5Sharp_tests/CMenu.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Document
{
    internal class CMenu
    {

        private TextReader _input { get; set; }
        private TextWriter _output { get; set; }
        private List<Item> _items = new List<Item>();
        private bool _exit = false;

        public CMenu(TextReader input, TextWriter output)
        {
            _input = input;
            _output = output;
        }

        public void AddItem(string shortcut, string description, ICommand command)
        {
            _items.Add(new Item(shortcut, description, command));
        }

        public void Run()
        {
            ShowInstructions();
            do
            {
                try
                {
                    _output.Write('>');
                    ExecuteCommand(_input.ReadLine()?.Split(" "));
                }
                catch (Exception ex)
                {   
                    _output.WriteLine(ex.Message);
                }
            } while (!_exit);
        }

        bool ExecuteCommand(string[]? args)
        {
            if (args == null)
                return !_exit;

            args = args.Where(x => !string.IsNullOrWhiteSpace(x)).ToArray();

            string command = args[0];

            int itemNum = _items.FindIndex(item => item.Shortcut == command);

            if (itemNum != -1)
            {
                _items[itemNum].Command.Execute(args.Skip(1).ToArray());
            }
            else
            {
                _output.WriteLine("Uncknown command");
            }

            return !_exit;
        }

        public void Exit()
        {
            _exit = true;
        }

        public void ShowInstructions()
        {
            _output.WriteLine("Command list:");
            _items.ForEach(item => _output.WriteLine($"\t{item.Shortcut}: {item.Description}"));
        }

        private struct Item
        {
            public Item(string shortcut, string description, ICommand command)
            {
                Shortcut = shortcut;
                Description = description;
                Command = command;
            }

            public string Shortcut { get; set; }
            public string Description { get; set; }
            public ICommand Command { get; set; }
        }
    }
}

File: Lw8/Task2/CMenu.cpp
#include "CMenu.h"

CMenu::CMenu(std::istream& in, std::ostream& out)
	: m_in(in)
	, m_out(out)
{
}

void CMenu::AddItem(const std::string& shortcut, const std::string& description, const Command& command)
{
	m_items.emplace_back(Item(shortcut, description, command));
}

void CMenu::Run()
{
	ShowInstructions();
	std::string command;
	while ((m_out << ">")
		&& m_in >> command
		&& ExecuteCommand(command))
	{
	}
}

void CMenu::ShowInstructions() const
{
	m_out << "Command list:\n";
	for (const auto& item : m_items)
	{
		m_out << "\t" << item.shortcut << ":\t" << item.description << "\n";
	}
}

void CMenu::Exit()
{
	m_exit = true;
}

bool CMenu::ExecuteCommand(const std::string& command)
{
	m_exit = false;

	auto it = std::find_if(m_items.begin(), m_items.end(), [&](const Item& item) {
		return item.shortcut == command;
	});

	if (it != m_items.end())
	{
		try
		{
			it->command();
		}
		catch (const std::exception& e)
		{
			m_out << e.what() << std::endl;
		}
	}
	else
	{
		m_out << "Uncknown command\n";
	}
	return !m_exit;
}
